Refactor surface conditions (#4532)
CliMA/ClimaAtmos.jl: docs/src/surface_conditions.md

# Surface Conditions

The lower boundary is where the atmosphere exchanges momentum, heat, and
moisture with whatever lies beneath it (ocean, land, sea ice, or an idealized
slab). ClimaAtmos collects everything controlling this boundary into one object,

> `AtmosSurface`, stored as `atmos.surface`. Each step it is read to fill

Reviewer: ```suggestion
`AtmosSurface`, stored as `atmos.surface`. It is read at each step to fill
```

`p.precomputed.sfc_conditions`, the surface fluxes and values consumed as
boundary conditions by the dynamical core, radiation, and turbulence schemes.

The [User Guide](#User-Guide) covers the options and how to choose; the
[Developer Guide](#Developer-Guide) covers the design, data flow, and how to
extend or debug it.

---

## User Guide

### The four knobs

`AtmosSurface` has four fields, each with one purpose:

> - `flux_scheme`: computes turbulent fluxes from the air–surface difference.

Reviewer: ```suggestion
- `flux_scheme`: computes turbulent fluxes from air–surface differences in temperature, humidity etc.
```

- `temperature`: sets the surface temperature `T_sfc`.

> - `boundary_overrides`: pins surface values that otherwise come from physics.

Reviewer: ```suggestion
- `boundary_overrides`: pins surface properties at user-specified values.
```

> - `surface_albedo`: sets the shortwave reflectivity seen by radiation.

Reviewer: Can albedo have a spectral dependence (e.g., near-infrared, visible, or direct radiation, diffuse radiation)?

Author: I added more detail to answer these questions in the `Albedo` section starting on line 100.

```@docs
ClimaAtmos.AtmosSurface
```

Set these directly when building a model, or let them be chosen by a
[setup](@ref "Setups") or by [YAML keys](#Configuring-from-YAML).

### Flux scheme (`flux_scheme`)

The closure turning the surface–to–lowest-level difference into turbulent
fluxes of momentum, heat, and moisture:

> - **`MoninObukhov`**: Monin–Obukhov Similarity Theory; fluxes follow from

Reviewer: ```suggestion
- **`MoninObukhov`**: Monin–Obukhov Similarity Theory (MOST); fluxes follow from
```

  roughness length and near-surface stability. Heat fluxes (`shf`/`lhf` or
  `θ_flux`/`q_flux`) or `ustar` may instead be prescribed (common for LES).
- **`ExchangeCoefficients`**: bulk fluxes with fixed `Cd`/`Ch`; simpler and

>   cheaper, for idealized constant exchange.

Reviewer: ```suggestion
  cheaper, for idealized constant exchange coefficients (rather than coefficients determined by MOST).
```

> - **`TimeVaryingMoninObukhov`**: wraps `f(t, FT)` returning a `MoninObukhov`

Reviewer: Does this mean the prescribed surface properties (e.g., surface temperature) vary in time? If so, it's a bit misleading to call it `TimeVaryingMoninObukhov`. It would rather be `TimeVaryingSurfaceProperties` (or something like this), which would be used with MOST or constant exchange coefficients. (And it would be a question whether this option is needed at all, or whether the time variation can be captured in the prescription of T_sfc and other surface properties.)

Author: This struct was just used for TRMM LBA and varied only shf/lhf, as time-varying surface temperature is already supported. I removed it, as we can allow the surface fluxes to vary with time by passing a function into `MoninObukhov(;fluxes = trmm_lba_fluxes)`

Author: On second thought, this might be unclear as we're just passing prescribed fluxes into MOST. I think this would be clarified by adding a separate `PrescribedSurfaceFluxes` type

Reviewer 2: It's MOST because momentum fluxes are still calculated with MOST. I think it's ok to leave it in MOST, otherwise we need to call surface_fluxes in PrescribedSurfaceFluxes too?

> - **`TimeVaryingMoninObukhov`**: wraps `f(t, FT)` returning a `MoninObukhov`

Reviewer 3: Can you make these clickable, e.g.
```
- **[`MoninObukhov`](@ref)**
```
or something similar

  each update; use when the flux forcing varies in time (e.g. TRMM_LBA).
- **`nothing`**: no atmos-side computation; an external driver supplies the
  conditions (see [Coupling](#Coupling-to-an-external-driver)).

```@docs
ClimaAtmos.SurfaceConditions.MoninObukhov
ClimaAtmos.SurfaceConditions.ExchangeCoefficients
ClimaAtmos.SurfaceConditions.TimeVaryingMoninObukhov
ClimaAtmos.SurfaceConditions.HeatFluxes
ClimaAtmos.SurfaceConditions.θAndQFluxes
```

### Temperature source (`temperature`)

What `T_sfc` is; the flux scheme then uses it (and surface humidity) for the
air–surface gradients:

- **`AnalyticTemperature(f)`**: `T_sfc = f(coordinates, params, t)`, per point.
  Covers a uniform constant (`AnalyticTemperature(Returns(FT(300)))`), a

>   zonally-symmetric SST, or a time-varying profile (e.g. GABLS).

Reviewer: ```suggestion
  zonally-symmetric SST, or a time-varying profile (e.g., GABLS).
```

- **`SlabOceanTemperature`**: *prognostic*; `T_sfc` read from `Y.sfc.T`, evolved

>   by a slab-ocean energy budget. The only type that adds prognostic state.

Reviewer: ```suggestion
  by a slab-ocean energy budget. The only type that adds a prognostic state.
```

```

!!! note "Constant temperature"
    There is no dedicated constant type. Use
    `AnalyticTemperature(Returns(FT(300)))`, wrapping the value in `FT(...)` to
    keep the broadcast type-stable.

### Boundary overrides (`boundary_overrides`)

By default, surface values come from physics (pressure hydrostatically
extrapolated, humidity saturated at `T_sfc`, zero winds, unit gustiness/moisture
availability). `SurfaceBoundaryOverrides` pins any of these to a fixed value;
each field defaults to `nothing` (use the physical default). Most idealized LES
setups override `p` and `q_vap`.

```@docs
ClimaAtmos.SurfaceConditions.SurfaceBoundaryOverrides
```

### Albedo (`surface_albedo`)

Sets the shortwave reflectivity seen by radiation: `ConstantAlbedo`,
`RegressionFunctionAlbedo`, or `CouplerAlbedo`. See the
[Ocean Surface Albedo](@ref "Ocean Surface Albedo") page.

### Choosing

| If you want… | `flux_scheme` | `temperature` |
|---|---|---|
| Stability-dependent fluxes over a prescribed SST | `MoninObukhov(; z0 = …)` | `AnalyticTemperature(…)` |
| Fixed-coefficient bulk fluxes | `ExchangeCoefficients(; Cd, Ch)` | `AnalyticTemperature(…)` |

> | Prescribed (possibly time-varying) heat fluxes | `MoninObukhov(; z0, shf, lhf)` / `TimeVaryingMoninObukhov` | `AnalyticTemperature(Returns(…))` |

Reviewer: I'm confused by this. If we use prescribed heat fluxes, we do not need MOST: the fluxes are just prescribed, independent of any MOST calculation of exchange coefficients.

> | Prescribed (possibly time-varying) heat fluxes | `MoninObukhov(; z0, shf, lhf)` / `TimeVaryingMoninObukhov` | `AnalyticTemperature(Returns(…))` |

Reviewer 3: ```suggestion
| Prescribed (possibly time-varying) heat fluxes | `MoninObukhov(; z0, shf, lhf)` / `TimeVaryingMoninObukhov` | `AnalyticTemperature(…)` |
```

> | An interactive surface temperature | `MoninObukhov(…)` | `SlabOceanTemperature(…)` |

Reviewer: ```suggestion
| An interactive slab ocean surface | `MoninObukhov(…)` | `SlabOceanTemperature(…)` |
```

| Surface temperature from data | `MoninObukhov(…)` | `ExternalTemperature()` |
| The coupler to own the surface | `nothing` or `MoninObukhov(…)` | `CoupledTemperature(field)` |

### Configuring from YAML

Three keys map onto `AtmosSurface` (resolved by
`AtmosSurface(::AtmosConfig, params, FT; setup_type)`); setup-provided pieces
take precedence over these defaults:

- `surface_setup` sets `flux_scheme`: `"DefaultExchangeCoefficients"` (default),
  `"DefaultMoninObukhov"`, or `"PrescribedSurface"` (→ `nothing`).
- `prognostic_surface` sets `temperature`: `"PrescribedSST"` (default) or
  `"SlabOceanSST"` (→ `SlabOceanTemperature`).
- `albedo_model` sets `surface_albedo`: `"ConstantAlbedo"` (default),

>   `"RegressionFunctionAlbedo"`, or `"CouplerAlbedo"`.

Reviewer 3: Could you add links to the doc sections for `flux_scheme`, `temperature`, `surface_albedo`.

Also missing a config option for `boundary_overrides`, I think

```@docs
ClimaAtmos.SurfaceConditions.DefaultMoninObukhov
ClimaAtmos.SurfaceConditions.DefaultExchangeCoefficients

> ```

Reviewer 3: These functions aren't really contextualized; adding a sentence introducing this list of methods would be useful. It also seems to be incomplete (e.g. missing some types that appear in the lists above)

### Coupling to an external driver

Under ClimaCoupler the coupler owns the surface, via one of two patterns:

1. **`flux_scheme = nothing`** (YAML `"PrescribedSurface"`): the atmosphere skips
   surface computation. `update_surface_conditions!` early-returns, and
   `init_sfc_conditions_zero!` pre-fills safe defaults at cache-build so RRTMGP /
   diagnostic EDMF never see uninitialized memory; the coupler then overwrites
   `sfc_conditions`.
2. **`CoupledTemperature(field)`**: the atmosphere still computes fluxes but
   reads `T_sfc` from a coupler-written `Field`. Per-cell overrides can be a
   `Fields.Field{<:SurfaceBoundaryOverrides}` on the cache. See

>    `test/coupler_compatibility.jl`.

Reviewer 3: I'm a bit confused by this section. The "Choosing" section above indicates that you should set _both_ `flux_scheme` and `temperature`. But this reads as if you only need to do one of these?

---

## Developer Guide

### Design: one source of truth

Surface behavior, once spread across a `surface_setup` kwarg, an `AbstractSST`

> hierarchy, and model fields, now lives entirely on `atmos.surface`. Principles:

Reviewer: ```suggestion
Surface behavior lives entirely on `atmos.surface`. Principles:
```

Reviewer: (No need to refer to past code versions.)

- **Orthogonality**: `flux_scheme`, `temperature`, `boundary_overrides`, and
  `surface_albedo` are independent axes. Adding an option on one shouldn't touch
  the others.
- **Dispatch over branching**: behavior is selected by dispatch on concrete
  types, not `if/elseif` on config strings.
- **Eager resolution**: YAML markers and `Default*` placeholders resolve to
  concrete structs at construction, so the hot path sees only concrete types.

### Data flow

The entry point `update_surface_conditions!` (called from
`set_precomputed_quantities!`) does four things: (1) early-return if
`isnothing(flux_scheme)`; (2) resolve the temperature via `surface_temperature`;
(3) resolve the flux scheme via `resolve_flux_scheme` (once per update);

> (4) broadcast `surface_state_to_conditions` over every surface point.

Reviewer 3: can you make these functions links? + various instances below

availability).
[`SurfaceBoundaryOverrides`](@ref ClimaAtmos.SurfaceConditions.SurfaceBoundaryOverrides)
pins any of these to a fixed value;
each field defaults to `nothing` (use the physical default). Most idealized LES
setups override `p` and `q_vap`.

### Albedo (`surface_albedo`)

Sets the shortwave reflectivity passed to the radiation scheme. Three models:

- **[`ConstantAlbedo`](@ref ClimaAtmos.ConstantAlbedo)**: a single value applied
  to both direct and diffuse shortwave.
- **[`RegressionFunctionAlbedo`](@ref ClimaAtmos.RegressionFunctionAlbedo)**: the
  Jin et al. (2011) ocean parameterization — a solar-zenith-angle-dependent
  *direct* albedo plus a separate *diffuse* albedo, with wind-speed-dependent
  surface roughness.
- **[`CouplerAlbedo`](@ref ClimaAtmos.CouplerAlbedo)**: albedo supplied by an
  external driver (the coupler).

**Direct vs. diffuse** The model carries distinct
`direct_sw_surface_albedo` and `diffuse_sw_surface_albedo` fields.
`ConstantAlbedo` sets them equal, `RegressionFunctionAlbedo` computes them
separately.

**Spectral** All three models write a single value across every shortwave band, and the

> Jin scheme treats the refractive index as wavelength-independent. The RRTMGP

Reviewer 3: ```suggestion
[`RegressionFunctionAlbedo`](@ref ClimaAtmos.RegressionFunctionAlbedo) scheme treats the refractive index as wavelength-independent. The RRTMGP
```

interface arrays are band-resolved (`(nbnd_sw, ncol)`), so per-band albedo is a
supported extension point but it would require a model that fills bands with distinct values.

**Longwave surface reflectivity** Albedo is shortwave-only, longwave surface reflectivity is handled
separately through `surface_emissivity`. 

See the [Ocean Surface Albedo](@ref "Ocean Surface Albedo") page for the

> Jin (2011) formulation.

Reviewer 3: ```suggestion
Jin (2011) [`RegressionFunctionAlbedo`](@ref ClimaAtmos.RegressionFunctionAlbedo) formulation.
```

### Choosing

`flux_scheme` and `temperature` are independent axes, and you set **both** (the
other two fields take defaults). Each row below is a compatible *pair*, not an
either/or:

| If you want… | `flux_scheme` | `temperature` |
|---|---|---|
| Stability-dependent fluxes over a prescribed SST | `MoninObukhov(; z0 = …)` | `AnalyticTemperature(…)` |
| Fixed-coefficient bulk fluxes | `ExchangeCoefficients(; Cd, Ch)` | `AnalyticTemperature(…)` |
| Prescribed heat fluxes (constant or time-varying) | `MoninObukhov(; z0, shf, lhf)` or `MoninObukhov(; z0, fluxes = (t,FT)->…)` | `AnalyticTemperature(…)` |
| An interactive slab ocean surface | `MoninObukhov(…)` | `SlabOceanTemperature(…)` |
| Surface temperature from data | `MoninObukhov(…)` | `ExternalTemperature()` |
| Coupler owns the surface (atmos skips fluxes) | `nothing` | unused — coupler writes `sfc_conditions` |

> | Coupler sets SST; atmos computes fluxes | `MoninObukhov(…)` | `CoupledTemperature(field)` |

Reviewer 3: ```suggestion
| Stability-dependent fluxes over a prescribed SST  | [`MoninObukhov(; z0 = …)`](@ref ClimaAtmos.SurfaceConditions.MoninObukhov)                    | [`AnalyticTemperature(…)`](@ref ClimaAtmos.SurfaceConditions.AnalyticTemperature)                                                                                     |
| Fixed-coefficient bulk fluxes                     | [`ExchangeCoefficients(; Cd, Ch)`](@ref ClimaAtmos.SurfaceConditions.ExchangeCoefficients)    | [`AnalyticTemperature(…)`](@ref ClimaAtmos.SurfaceConditions.AnalyticTemperature)                                                                                     |
| Prescribed heat fluxes (constant or time-varying) | [`MoninObukhov(; z0, shf, lhf)`](@ref ClimaAtmos.SurfaceConditions.MoninObukhov) or [`MoninObukhov(; z0, fluxes = (t,FT)->…)`](@ref ClimaAtmos.SurfaceConditions.MoninObukhov) | [`AnalyticTemperature(…)`](@ref ClimaAtmos.SurfaceConditions.AnalyticTemperature) |
| An interactive slab ocean surface                 | [`MoninObukhov(…)`](@ref ClimaAtmos.SurfaceConditions.MoninObukhov)                           | [`SlabOceanTemperature(…)`](@ref ClimaAtmos.SurfaceConditions.SlabOceanTemperature)                                                                                   |
| Surface temperature from data                     | [`MoninObukhov(…)`](@ref ClimaAtmos.SurfaceConditions.MoninObukhov)                           | [`ExternalTemperature(…)`](@ref ClimaAtmos.SurfaceConditions.ExternalTemperature)                                                                                     |
| Coupler owns the surface (atmos skips fluxes)     | `nothing`                                                                                     | unused — coupler writes `sfc_conditions`                                                                                                                              |
| Coupler sets SST; atmos computes fluxes           | [`MoninObukhov(…)`](@ref ClimaAtmos.SurfaceConditions.MoninObukhov)                           | [`CoupledTemperature(field)`](@ref ClimaAtmos.SurfaceConditions.CoupledTemperature)                                                                                   |
```
